coderd/workspaceagentsrpc.go

```diff
@@ -184,73 +166,137 @@ func ensureLatestBuild(ctx context.Context, db database.Store, logger slog.Logge
 			Message: "Agent trying to connect from non-latest build.",
 			Detail:  err.Error(),
 		})
-		return nil, database.WorkspaceBuild{}, false
+		return database.WorkspaceBuild{}, false
 	}
 
-	return ensureLatestBuild, build, true
+	return build, true
 }
 
-func (api *API) agentConnectionUpdate(ctx context.Context, workspaceAgent database.WorkspaceAgent, workspaceID uuid.UUID, conn *websocket.Conn) (func(closeCtx context.Context, ensureLatestBuildFn func() error), bool) {
-	// We use a custom heartbeat routine here instead of `httpapi.Heartbeat`
-	// because we want to log the agent's last ping time.
-	var lastPing atomic.Pointer[time.Time]
-	lastPing.Store(ptr.Ref(time.Now())) // Since the agent initiated the request, assume it's alive.
-
-	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), func(ctx context.Context) {
-		// TODO(mafredri): Is this too frequent? Use separate ping disconnect timeout?
-		t := time.NewTicker(api.AgentConnectionUpdateFrequency)
-		defer t.Stop()
-
-		for {
-			select {
-			case <-t.C:
-			case <-ctx.Done():
-				return
-			}
+func checkBuildIsLatest(ctx context.Context, db database.Store, build database.WorkspaceBuild) error {
+	latestBuild, err := db.GetLatestWorkspaceBuildByWorkspaceID(ctx, build.WorkspaceID)
+	if err != nil {
+		return err
+	}
+	if build.ID != latestBuild.ID {
+		return xerrors.New("build is outdated")
+	}
+	return nil
+}
 
-			// We don't need a context that times out here because the ping will
-			// eventually go through. If the context times out, then other
-			// websocket read operations will receive an error, obfuscating the
-			// actual problem.
-			err := conn.Ping(ctx)
-			if err != nil {
-				return
-			}
-			lastPing.Store(ptr.Ref(time.Now()))
+func (api *API) agentConnectionUpdate(ctx context.Context,
+	workspaceAgent database.WorkspaceAgent, workspaceBuild database.WorkspaceBuild,
+	conn *websocket.Conn,
+) {
+	minder := &agentWebsocketMinder{
+		apiCtx:            api.ctx,
+		workspaceAgent:    workspaceAgent,
+		workspaceBuild:    workspaceBuild,
+		conn:              conn,
+		pingPeriod:        api.AgentConnectionUpdateFrequency,
+		db:                api.Database,
+		replicaID:         api.ID,
+		updater:           api,
+		disconnectTimeout: api.AgentInactiveDisconnectTimeout,
+		logger:            api.Logger.With(slog.F("workspace_id", workspaceBuild.WorkspaceID)),
+	}
+	minder.init()
+	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), minder.sendPings)
+	go pprof.Do(ctx, pprof.Labels("agent", workspaceAgent.ID.String()), minder.mind)
```

Comment:
Would be nice to have a Close so we can ensure goroutines are cleaned up before we return.

Reply:
These routines take a context and complete when the context is done. I don't see the point of blocking the HTTP handler on closing down these goroutines; what would it accomplish?

Reply:
It ensures any finalizers (if any) are run before process exit and helps reason about the lifecycle of the service. Contexts tell us nothing except “I’ll stop executing sometime in the future”.

Reply:
That's a good point --- we have websocket tracking in Coderd, so it would be nice to ensure the db update that marks the agent disconnected gets a chance to be written.